README.md
# Brainfoose

Yet another brainfuck interperator.

### Usage

The following characters make up the language:

| Character | Meaning                                                          |
| --------- | ---------------------------------------------------------------- |
| `>`       | Increment the data pointer.                                      |
| `<`       | Decrement the data pointer.                                      |
| `+`       | Increment the value at the data pointer.                         |
| `-`       | Decrement the value at the data pointer.                         |
| `.`       | Output the byte at the data pointer.                             |
| `,`       | Store a value at the data pointer's position.                    |
| `[`       | If the value at the data pointer is 0, move to matching `]`.     |
| `]`       | If the value at the data pointer is not 0, more to matching `[`. |
| `&`       | Reset array.                                                     |
| `%`       | Load program from file, syntax: `% [filename.bf]`.               |

To quit the prompt press <kbd>ctrl</kbd> <kbd>c</kbd>.

[![demo](https://asciinema.org/a/9ecaa1s4wj8rctm0hypzrg595.png)](https://asciinema.org/a/9ecaa1s4wj8rctm0hypzrg595?autoplay=1)

The following is hello world program, taken from [wikipedia][wbfe].

```brainfuck
[ This program prints "Hello World!" and a newline to the screen, its
  length is 106 active command characters. [It is not the shortest.]

  This loop is an "initial comment loop", a simple way of adding a comment
  to a BF program such that you don't have to worry about any command
  characters. Any ".", ",", "+", "-", "<" and ">" characters are simply
  ignored, the "[" and "]" characters just have to be balanced. This
  loop and the commands it contains are ignored because the current cell
  defaults to a value of 0; the 0 value causes this loop to be skipped.
]
++++++++               Set Cell #0 to 8
[
    >++++               Add 4 to Cell #1; this will always set Cell #1 to 4
    [                   as the cell will be cleared by the loop
        >++             Add 2 to Cell #2
        >+++            Add 3 to Cell #3
        >+++            Add 3 to Cell #4
        >+              Add 1 to Cell #5
        <<<<-           Decrement the loop counter in Cell #1
    ]                   Loop till Cell #1 is zero; number of iterations is 4
    >+                  Add 1 to Cell #2
    >+                  Add 1 to Cell #3
    >-                  Subtract 1 from Cell #4
    >>+                 Add 1 to Cell #6
    [<]                 Move back to the first zero cell you find; this will
                        be Cell #1 which was cleared by the previous loop
    <-                  Decrement the loop Counter in Cell #0
]                       Loop till Cell #0 is zero; number of iterations is 8

The result of this is:
Cell No :   0   1   2   3   4   5   6
Contents:   0   0  72 104  88  32   8
Pointer :   ^

>>.                     Cell #2 has value 72 which is 'H'
>---.                   Subtract 3 from Cell #3 to get 101 which is 'e'
+++++++..+++.           Likewise for 'llo' from Cell #3
>>.                     Cell #5 is 32 for the space
<-.                     Subtract 1 from Cell #4 for 87 to give a 'W'
<.                      Cell #3 was set to 'o' from the end of 'Hello'
+++.------.--------.    Cell #3 for 'rl' and 'd'
>>+.                    Add 1 to Cell #5 gives us an exclamation point
>++.                    And finally a newline from Cell #6
```

This code can also be rewritten without comments as:

```brainfuck
++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.
```

[wbfe]: https://en.wikipedia.org/wiki/Brainfuck#Hello_World.21


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. The format
is based on [Keep a Changelog](http://keepachangelog.com/) and this project
adheres to [Semantic Versioning](http://semver.org/).

## [2.0.1] - 2017-10-31
### Changed
-   Entering `?` now displays a help message.

## [2.0.0] - 2017-08-14
### Changed
-   All code has been refactored.

## [1.0.0] - 2017-03-15
### Added
-   Initial working release

[2.0.0]: https://github.com/abactel/brainfoose/compare/v2.0.0...v2.0.1
[2.0.0]: https://github.com/abactel/brainfoose/compare/v1.0.0...v2.0.0
[1.0.0]: https://github.com/abactel/brainfoose/releases/tag/V1.0.0


tests/test.py
#!/usr/bin/env python3

import brainfoose
import unittest

from termcolor import colored

def read_file(filename:str):
    with open(filename) as file:
        return file.readline()

class TestExecution(unittest.TestCase):

    def setUp(self):
        pass

    # basic stuff
    def test_create_tape(self):
        tape = brainfoose.new_tape()
        correct = [0] * 3000
        self.assertEqual(tape, correct)

    def test_create_tape_5_cells(self):
        tape = brainfoose.new_tape(5)
        correct = [0] * 5
        self.assertEqual(tape, correct)

    def test_add_10_to_first_cell(self):
        tape = brainfoose.new_tape(tape_size=5)
        bf, out = brainfoose.execute_program("++++++++++", tape)
        correct = [10, 0, 0, 0, 0]
        self.assertEqual(bf, correct)

    def test_add_5_to_first_two_cells(self):
        tape = brainfoose.new_tape(5)
        bf, out = brainfoose.execute_program("+++++>+++++", tape)
        correct = [5, 5, 0, 0, 0]
        self.assertEqual(bf, correct)

    def test_add_5_subtract_4(self):
        tape = brainfoose.new_tape(5)
        bf, out = brainfoose.execute_program("+++++----", tape)
        correct = [1, 0, 0, 0, 0]
        self.assertEqual(bf, correct)

    def test_add_2_to_all_cells(self):
        tape = brainfoose.new_tape(5)
        bf, out = brainfoose.execute_program("++>++>++>++>++", tape)
        correct = [2, 2, 2, 2, 2]
        self.assertEqual(bf, correct)

    def test_add_2_to_all_cells_then_take_away_1(self):
        tape = brainfoose.new_tape(5)
        bf, out = brainfoose.execute_program("++>++>++>++>++<-<-<-<-<-", tape)
        correct = [1, 1, 1, 1, 1]
        self.assertEqual(bf, correct)

    def test_add_first_two_cells(self):
        tape = brainfoose.new_tape(5)
        bf, out = brainfoose.execute_program("++>++<[->+<]", tape)
        correct = [0, 4, 0, 0, 0]
        self.assertEqual(bf, correct)


    # errors
    def test_error_if_unequal_brackets(self):
        tape = brainfoose.new_tape(5)
        bf, out = brainfoose.execute_program("[", tape)
        correct = colored("Brackets don't match.", "red")
        self.assertEqual(out, correct)

    def test_error_if_tape_too_small(self):
        tape = brainfoose.new_tape(5)
        bf, out = brainfoose.execute_program(">>>>>+", tape)
        correct = colored("Not enough tape.", "red")
        self.assertEqual(out, correct)

    # misc
    def test_hello_world(self):
        tape = brainfoose.new_tape()
        bf, out = brainfoose.execute_program(read_file("hello_world.txt"), tape)
        correct = "Hello World!\n"
        self.assertEqual(out, correct)


if __name__ == '__main__':
    unittest.main()


brainfoose.py
#!/usr/bin/env python3

import sys  # Allows exiting

BOLD = '\033[1m'
END_BOLD = '\033[0m'

def main():
    def new_tape():
        return [0] * 30000
    print(BOLD + "\nBrainfoose REPL" + END_BOLD)
    tape = new_tape()
    while True:
        try:
            program = input(BOLD + "bf > " + END_BOLD)
            if program[0] == "%":
                with open(program[2:], "r") as file:
                    program = ""
                    for line in file.readlines():
                        program += line
        except KeyboardInterrupt:
            print("\nBye.")
            sys.exit(0)
        end_index = len(program) - 1
        pointer = 0
        program_location = 0
        while program_location <= end_index:
            token = program[program_location]
            if token == ">":
                pointer += 1
            elif token == "<":
                pointer -= 1
            elif token == "+":
                tape[pointer] += 1
            elif token == "-":
                tape[pointer] -= 1
            elif token == ".":
                print(chr(tape[pointer]), end="")
            elif token == ",":
                tape[pointer] = ord(input(BOLD + "input > " + END_BOLD)[0])
            elif token == "[":
                if tape[pointer] == 0:
                    loop_level = 1
                    while loop_level > 0:
                        program_location += 1
                        character = program[program_location]
                        if character == "[":
                            loop_level += 1
                        elif character == "]":
                            loop_level -= 1
            elif token == "]":
                loop_level = 1
                while loop_level > 0:
                    program_location -= 1
                    character = program[program_location]
                    if character == "[":
                        loop_level -= 1
                    elif character == "]":
                        loop_level += 1
                program_location -= 1
            elif token == "&":
                tape = new_tape()
            program_location += 1


if __name__ == '__main__':
    main()


braingoose.py
#!/usr/bin/env python3

BOLD = '\033[1m'
END_BOLD = '\033[0m'

def main():
    print(BOLD + "\nBraingoose REPL" + END_BOLD)
    array = [0] * 30000
    while True:
        program = input("\nbg > ")
        end_index = len(program) - 1
        pointer = 0
        program_location = 0

        while program_location <= end_index:
            token = program[program_location]
            if token == ">":
                pointer += 1
            elif token == "<":
                pointer -= 1
            elif token == "+":
                array[pointer] += 1
            elif token == "-":
                array[pointer] -= 1
            elif token == ".":
                print(chr(array[pointer]), end="")
            elif token == ",":
                array[pointer] = ord(input("input > ")[0])
            elif token == "[":
                if array[pointer] == 0:
                    loop_level = 1
                    while loop_level > 0:
                        program_location += 1
                        character = program[program_location]
                        if character == "[":
                            loop_level += 1
                        elif character == "]":
                            loop_level -= 1
            elif token == "]":
                loop_level = 1
                while loop_level > 0:
                    program_location -= 1
                    character = program[program_location]
                    if character == "[":
                        loop_level -= 1
                    elif character == "]":
                        loop_level += 1
                program_location -= 1
            elif token == "&":
                array = [0] * 30000
            program_location += 1
    print("\n")


if __name__ == '__main__':
    main()


brainfoose/__init__.py
#!/usr/bin/env python3

"""Brainfoose.

A brainfuck REPL.

Usage:
    brainfoose [--run_command=<programs>]
    brainfoose [--tape_size=<tape_size>]
    brainfoose [--verify_program=<tape_size>]
    brainfoose (-h | --help)
    brainfoose --version

Options:
    -h, --help               Show this message
    --version                Show version information
    --run                    Run a command without starting a REPL
    --tape_size=<tape_size>  Set size of tape [default: 3000]

REPL syntax:
    ?  Print this help screen
    >  Increment the data pointer by 1
    <  Decrement the data pointer by 1
    +  Increment the value at the data pointer by 1
    -  Decrement the value at the data pointer by 1
    .  Output the byte at the data pointer. The byte is formatted to the
         character it represents
    ,  Store a value at the data pointer’s position. The character is as a
         decimal byte representing the character. The character is accepted
         form a standard input prompt
    [  If the value at the data pointer is 0, move to matching ]
    ]  If the value at the data pointer is not 0, move to matching [
    $  Display tape up to last non-zero cell
    &  Reset tape
    %  Load program from file, syntax: %[filename]

"""


from docopt import docopt      # Command line argument parsing
from termcolor import colored  # Color output


def get_input(text: str):
    """Handle user input.

    The function exits cleanly on ``KeyboardInterrupt`` and ``EOFError``
    (ctrl-c and ctrl-d).

    Args:
        Text (str): Prompt to print. A greater than symbol (">") is appended to
            the end and escape codes are used to make the prompt bold.

    Returns:
        Str: User's input.

    """
    try:
        user_input = input("\033[1m" + text + " > \033[0m")
    except (KeyboardInterrupt, EOFError):
        print("\nBye.")
        exit(0)
    return user_input


def get_program():
    """Gets user input and opens program file if needed.

    Returns:
        Str: program

    """
    program = get_input("bf") + " "  # space need to prevent out of index error
    if program.startswith("%"):
        try:
            with open(program[1:], "r") as file:
                program = ""
                for line in file.readlines():
                    program += line
        except FileNotFoundError:
            print("File not found: " + program[2:])
            program = ""
    elif program.startswith("?"):
        print(__doc__)
    return program


def new_tape(tape_size=3000):
    """Creates a new tape.

    Args:
        Tape_size (int) (optional): Size of the tape [default:3000].

    Returns:
        list: Tape. A list of integer ``0``s of specified size.

    """
    return [0] * tape_size


def execute_program(program: str, tape: list):
    """Runs the brainfuck program.

    Args:
        Program (str): A brainfuck program.
        Tape (list): A list of integers (see new_tape()).

    Returns:
        List: Tape (see above).
        Str: Output to print.

    """
    output = []
    if not program.count("[") == program.count("]"):
        output.append(colored("Brackets don't match.", "red"))
        return tape, "".join(output)
    if program.count(">") >= len(tape):
        output.append(colored("Not enough tape.", "red"))
        return tape, "".join(output)
    end_index = len(program) - 1
    pointer = 0
    program_location = 0
    while program_location <= end_index:
        token = program[program_location]
        print_at_end = ""  # adds final linebreak after printing chars
        if token == ">":
            pointer += 1
        elif token == "<":
            pointer -= 1
        elif token == "+":
            tape[pointer] += 1
        elif token == "-":
            tape[pointer] -= 1
        elif token == ".":
            output.append(chr(tape[pointer]))
        elif token == ",":
            output.append(print_at_end)
            print_at_end = ""
            tape[pointer] = ord(get_input("input")[0])
        elif token == "[":
            if tape[pointer] == 0:
                loop_level = 1
                while loop_level > 0:
                    program_location += 1
                    sub_token = program[program_location]
                    if sub_token == "[":
                        loop_level += 1
                    elif sub_token == "]":
                        loop_level -= 1
        elif token == "]":
            loop_level = 1
            while loop_level > 0:
                program_location -= 1
                sub_token = program[program_location]
                if sub_token == "[":
                    loop_level -= 1
                elif sub_token == "]":
                    loop_level += 1
            program_location -= 1
        elif token == "&":
            tape = new_tape()
        elif token == "$":
            last_important_value = 0
            for cell_n, cell in enumerate(tape):
                if cell != 0:
                    last_important_value = cell_n
            output.append(str(tape[:last_important_value + 1]))
        program_location += 1
    output.append(print_at_end)
    return tape, "".join(output)


def main():
    """Starts REPL or runs program."""
    arguments = docopt(__doc__, version="Brainfoose v2.0.0")
    tape = new_tape(int(arguments["--tape_size"]))
    print("\n\033[1mBrainfoose.\033[0m\n")  # brainfoose in bold plus newlines
    if arguments["--run_command"]:
        tape, output = execute_program(arguments["--run_command"], tape)
        print(output)
    else:
        while True:
            program = get_program()
            tape, output = execute_program(program, tape)
            print(output)


if __name__ == '__main__':
    main()
